File: LeetCode/python/104.py
# source : https://leetcode.com/problems/maximum-depth-of-binary-tree/

from typing import Optional
from collections import deque

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
        
class Solution:
    def maxDepth(self, root: Optional[TreeNode]) -> int:
        if not root:
            return 0
        
        depth = 0
        queue = deque([root])
        
        while queue:
            # level 당 +1
            depth += 1
            
            for _ in range(len(queue)):
                temp = queue.popleft()
                
                if temp.left:
                    queue.append(temp.left)
                
                if temp.right:
                    queue.append(temp.right)
        
        return depth

File: LeetCode/javascript/104.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */

/**
 * @param {TreeNode} root
 * @return {number}
 */
 var maxDepth = function(root) {
    const queue = [root];
    let depth = 0;
    
    if (root === null) {
        return 0;
    }
    
    while (queue.length) {
        depth++;
        const nodeCount = queue.length;
        
        for (i = 0; i < nodeCount; i++) {
            currNode = queue[0];
            
            if (currNode?.left) {
                queue.push(currNode.left);
            }
            if (currNode?.right) {
                queue.push(currNode.right);
            }

            queue.shift();
        }
    }
    
    return depth;
};